contracts/Controller.sol

```diff
@@ -245,46 +240,41 @@ contract Controller is IController {
         // VoterProxy so that it can grab the incentive tokens off the contract after claiming rewards
         // RewardFactory so that stashes can make new extra reward contracts if a new incentive is added to the gauge
         poolInfo[pid].stash = stash;
-        IVoterProxy(staker).grantStashAccess(stash);
         IRewardFactory(rewardFactory).grantRewardStashAccess(stash);
         redirectGaugeRewards(stash, _gauge);
         emit AddedPool(pid, _lptoken, token, _gauge, newRewardPool, stash);
     }
 
-    /// @notice shuts down a currently active pool
-    /// @param _pid The id of the pool to shutdown
-    function shutdownPool(uint256 _pid) external onlyAddress(poolManager) {
-        PoolInfo storage pool = poolInfo[_pid];
+    /// @notice Shuts down multiple pools
+    /// @dev Claims rewards for that pool before shutting it down
+    /// @param _startPoolIdx Start pool index
+    /// @param _endPoolIdx End pool index (excluded)
+    function bulkPoolShutdown(uint256 _startPoolIdx, uint256 _endPoolIdx) external onlyAddress(poolManager) {
+        for (uint256 i = _startPoolIdx; i < _endPoolIdx; ++i) {
+            PoolInfo storage pool = poolInfo[i];
+
+            if (pool.shutdown) {
+                continue;
+            }
+
+            _earmarkRewards(i);
 
-        //withdraw from gauge
-        // solhint-disable-next-line
-        try IVoterProxy(staker).withdrawAll(pool.lptoken, pool.gauge) {
+            //withdraw from gauge
             // solhint-disable-next-line
-        } catch {}
+            try IVoterProxy(staker).withdrawAll(pool.lptoken, pool.gauge) {
+                // solhint-disable-next-line
+            } catch {}
 
-        pool.shutdown = true;
-        gaugeMap[pool.gauge] = false;
-        emit PoolShutDown(_pid);
+            pool.shutdown = true;
+            gaugeMap[pool.gauge] = false;
+            emit PoolShutDown(i);
+        }
     }
 
     /// @notice shuts down all pools
-    /// @dev This shuts down the contract, unstakes and withdraws all LP tokens
+    /// @dev This shuts down the contract
     function shutdownSystem() external onlyAddress(owner) {
         isShutdown = true;
```

Comment:
System shutdown just does nothing except for changing bool?
What about unstaking and withdrawing tokens?

Reply:
Correct. In the previous implementation, we had a for loop which tries to withdraw tokens, and do a lot of stuff.
I went to [Tenderly](https://tenderly.co/) and I've simulated `shutdownSystem` on Convex and it reverted.
Here is a tx receipt, the Convex controller is called from the owner, and it reverted because it consumed all the gas.
![image](https://user-images.githubusercontent.com/28648109/177931841-d25577ed-f614-41e9-bb39-c77accd8394e.png) 
That is a design issue that I've resolved by splitting the functionality in two parts.
1. `bulkPoolShutdown` allows for poolManager to shut down multiple pools in a single transaction (on cvx he would need to make a lot of transactions because he can't use multi call because of authorization.
2. system shutdown (prevents deposits and new pool adding)

Reply:
Ooof, such a marvelous system
Yup, reasoning sounds solid, and as I see it can be called after the system is shutdown, so nice